main/src/main/scala/sbt/internal/server/BspCompileTask.scala

```diff
@@ -52,66 +51,85 @@ object BspCompileTask {
     BspCompileTask(targetId, targetName, taskId, System.currentTimeMillis())
   }
 
-  private def notifyStart(task: BspCompileTask): Unit = {
-    val message = s"Compiling ${task.targetName}"
-    val data = Converter.toJsonUnsafe(CompileTask(task.targetId))
-    val params = TaskStartParams(task.id, task.startTimeMillis, message, "compile-task", data)
+  private def compileReport(
+      problems: Seq[Problem],
+      targetId: BuildTargetIdentifier,
+      elapsedTimeMillis: Long
+  ): CompileReport = {
+    val countBySeverity = problems.groupBy(_.severity()).mapValues(_.size)
+    val warnings = countBySeverity.getOrElse(Severity.Warn, 0)
+    val errors = countBySeverity.getOrElse(Severity.Error, 0)
+    CompileReport(targetId, None, errors, warnings, Some(elapsedTimeMillis.toInt))
+  }
+}
+
+case class BspCompileTask private (
+    targetId: BuildTargetIdentifier,
+    targetName: String,
+    id: TaskId,
+    startTimeMillis: Long
+) {
+  import sbt.internal.bsp.codec.JsonProtocol._
+
+  private[sbt] def notifyStart(): Unit = {
+    val message = s"Compiling $targetName"
+    val data = Converter.toJsonUnsafe(CompileTask(targetId))
+    val params = TaskStartParams(id, startTimeMillis, message, "compile-task", data)
     exchange.notifyEvent("build/taskStart", params)
   }
 
-  private def notifySuccess(task: BspCompileTask, result: CompileResult): Unit = {
+  private[sbt] def notifySuccess(result: CompileResult): Unit = {
     import collection.JavaConverters._
     val endTimeMillis = System.currentTimeMillis()
-    val elapsedTimeMillis = endTimeMillis - task.startTimeMillis
+    val elapsedTimeMillis = endTimeMillis - startTimeMillis
     val problems = result match {
       case compileResult: CompileResult =>
         val sourceInfos = compileResult.analysis().readSourceInfos().getAllSourceInfos.asScala
         sourceInfos.values.flatMap(_.getReportedProblems).toSeq
       case _ => Seq()
     }
-    val report = compileReport(problems, task.targetId, elapsedTimeMillis)
+    val report = compileReport(problems, targetId, elapsedTimeMillis)
     val params = TaskFinishParams(
-      task.id,
+      id,
       endTimeMillis,
-      s"Compiled ${task.targetName}",
+      s"Compiled $targetName",
       StatusCode.Success,
       "compile-report",
       Converter.toJsonUnsafe(report)
     )
     exchange.notifyEvent("build/taskFinish", params)
   }
 
-  private def notifyFailure(task: BspCompileTask, cause: Option[CompileFailed]): Unit = {
+  private[sbt] def notifyProgress(percentage: Int, total: Int): Unit = {
+    val data = Converter.toJsonUnsafe(CompileTask(targetId))
+    val message = s"Compiling $targetName ($percentage%)"
+    val currentMillis = System.currentTimeMillis()
+    val params = TaskProgressParams(
+      id,
+      Some(currentMillis),
+      Some(message),
+      Some(total.toLong),
+      Some(percentage.toLong),
```

Comment:
unfortunately this should have been the "current" value, and not the percentage

Reply:
How does IntelliJ handle it? How does it handle `bloop-progress`? Can we do the same in Metals?

Reply:
It looks like IntelliJ ignores the kind field:
https://github.com/JetBrains/intellij-scala/blob/5100ec6250070db147f409af755e08a75873e1a3/bsp/src/org/jetbrains/bsp/project/BspTask.scala#L298

I can only see that it prints the message field in the build window

Reply:
Ok then I think we should fix this in sbt so that it is consistent with Bloop.

Reply:
ok but I still propose that we change the `kind` of the event, this way metals will always be correct, as it can handle the old kind the old way